go/engine/merkle_audit.go

```diff
@@ -0,0 +1,246 @@
+		Settings: MerkleAuditSettings,
+
+		testingMetaCh:     args.testingMetaCh,
+		testingRoundResCh: args.testingRoundResCh,
+	})
+	return &MerkleAudit{
+		Contextified: libkb.NewContextified(g),
+		args:         args,
+		task:         task,
+	}
+}
+
+// Name is the unique engine name.
+func (e *MerkleAudit) Name() string {
+	return "MerkleAudit"
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *MerkleAudit) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *MerkleAudit) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *MerkleAudit) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+// Run starts the engine.
+// Returns immediately, kicks off a background goroutine.
+func (e *MerkleAudit) Run(m libkb.MetaContext) (err error) {
+	return RunEngine2(m, e.task)
+}
+
+func (e *MerkleAudit) Shutdown() {
+	e.task.Shutdown()
+}
+
+// randSeqno picks a random number between [low, high).
+func randSeqno(lo keybase1.Seqno, hi keybase1.Seqno) (keybase1.Seqno, error) {
+	rangeBig := big.NewInt(int64(hi - lo))
+	n, err := rand.Int(rand.Reader, rangeBig)
+	if err != nil {
+		return keybase1.Seqno(0), err
+	}
+	return keybase1.Seqno(n.Int64()) + lo, nil
+}
+
+var merkleAuditKey = libkb.DbKey{
+	Typ: libkb.DBMerkleAudit,
+	Key: "root",
+}
+
+func lookupMerkleAuditRetry(m libkb.MetaContext) (*keybase1.Seqno, error) {
+	var state merkleAuditState
+	found, err := m.G().LocalDb.GetInto(&state, merkleAuditKey)
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		// Nothing found, no error
+		return nil, nil
+	}
+
+	// Can still be nil
+	return state.RetrySeqno, nil
+}
+
+func saveMerkleAuditState(m libkb.MetaContext, state merkleAuditState) error {
+	return m.G().LocalDb.PutObj(merkleAuditKey, nil, state)
+}
+
+func performMerkleAudit(m libkb.MetaContext, startSeqno keybase1.Seqno) error {
+	if m.G().ConnectivityMonitor.IsConnected(m.Ctx()) == libkb.ConnectivityMonitorNo {
+		m.CDebugf("MerkleAudit giving up offline")
+		return errAuditOffline
+	}
+
+	// Acquire the most recent merkle tree root
+	lastRoot := m.G().MerkleClient.LastRoot()
+	if lastRoot == nil {
+		m.CDebugf("MerkleAudit unable to retrieve the last root")
+		return errAuditNoLastRoot
+	}
+
+	// We can copy the pointer's value as it can only return nil if root == nil.
+	lastSeqno := *lastRoot.Seqno()
+
+	// Acquire the first root and calculate its hash
+	startRoot, err := m.G().MerkleClient.LookupRootAtSeqno(m, startSeqno)
+	if err != nil {
+		return err
+	}
+	startHash := startRoot.ShortHash()
+
+	// Traverse the merkle tree seqnos
+	currentSeqno := startSeqno + 1
+	step := 1
+	for {
+		// Proceed until the last known root
+		if currentSeqno > lastSeqno {
+			break
+		}
+
+		currentRoot, err := m.G().MerkleClient.LookupRootAtSeqno(m, currentSeqno)
+		if err != nil {
+			return err
+		}
+		currentHash := currentRoot.SkipToSeqno(startSeqno)
```

Comment:
I guess this might happen if the server does not provide skip sequence for the root, but maybe it would bail out somewhere else then. I don't know this code very well

Reply:
165 would error first, just checked it.

Reply:
... actually I need to recheck this again, considering another usecase.

Reply:
verifySkipSequence gets called in the fetcher, which checks if the skip sequence should be there. Honestly it seems to me that most of the auditor code I wrote is useless, considering that we already verify these things in another place. Maybe just pulling the root should be enough?

Reply:
Checked it anyways.